psyneulink/library/compositions/compositionrunner.py

```diff
@@ -228,6 +228,10 @@ def run_learning(self,
         # return results from last epoch
         results = self._composition.parameters.results.get(context)[-1 * num_epoch_results:]
 
+        # FIXME: This currently does not work for pnlvm.ExecutionMode.Auto, since we have no way of knowing what mode was taken
+        if hasattr(self._composition, "_writeback_parameters") and (execution_mode == pnlvm.ExecutionMode.Python):
```

Comment:
can this ever run anything else than Autodiff composition?

Reply:
Yes; PsyNeuLink has learning mechanisms (https://princetonuniversity.github.io/PsyNeuLink/LearningMechanism.html) that are able to be learned by calling `composition.learn`. We only need the writeback in the case of autodiff though, since we have a pytorch backend that maintains separate copies of the weights

Reply:
so `self._composition` can be a learning mechanism?

Reply:
it can be a non-autodiff compostion, with learning mechanisms inside of it

Reply:
what if there is another autodiff nested inside? can you call learn on the outer composition?

Reply:
ah, I see your point.
Yes, this is definitely an edge-case that was missed. Good catch!

Reply:
Fixed!